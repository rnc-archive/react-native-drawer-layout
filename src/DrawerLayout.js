// @flow
import React, { Component } from 'react';
import dismissKeyboard from 'react-native-dismiss-keyboard';
import {
  Animated,
  Dimensions,
  PanResponder,
  StyleSheet,
  TouchableWithoutFeedback,
  View,
  I18nManager,
} from 'react-native';

const MIN_SWIPE_DISTANCE = 3;
const DEVICE_WIDTH = parseFloat(Dimensions.get('window').width);
const THRESHOLD = DEVICE_WIDTH / 2;
const VX_MAX = 0.1;

const IDLE = 'Idle';
const DRAGGING = 'Dragging';
const SETTLING = 'Settling';

export type PropType = {
  children: any,
  drawerBackgroundColor?: string,
  drawerLockMode?: 'unlocked' | 'locked-closed' | 'locked-open',
  drawerPosition: 'left' | 'right',
  drawerWidth: number,
  keyboardDismissMode?: 'none' | 'on-drag',
  onDrawerClose?: Function,
  onDrawerOpen?: Function,
  onDrawerSlide?: Function,
  onDrawerStateChanged?: Function,
  renderNavigationView: () => any,
  statusBarBackgroundColor?: string,
  useNativeAnimations?: boolean,
  maxOverlayOpacity: number,
};

export type StateType = {
  accessibilityViewIsModal: boolean,
  drawerShown: boolean,
  openValue: any,
};

export type EventType = {
  stopPropagation: Function,
};

export type PanResponderEventType = {
  dx: number,
  dy: number,
  moveX: number,
  moveY: number,
  vx: number,
  vy: number,
};

export type DrawerMovementOptionType = {
  velocity?: number,
};

export default class DrawerLayout extends Component {
  props: PropType;
  state: StateType;
  _lastOpenValue: number;
  _panResponder: any;
  _isClosing: boolean;
  _closingAnchorValue: number;

  static defaultProps = {
    drawerWidth: 0,
    drawerPosition: 'left',
    useNativeAnimations: false,
    maxOverlayOpacity: 0.7,
  };

  static positions = {
    Left: 'left',
    Right: 'right',
  };

  constructor(props: PropType, context: any) {
    super(props, context);

    this.state = {
      accessibilityViewIsModal: false,
      drawerShown: false,
      openValue: new Animated.Value(0),
    };
  }

  getDrawerPosition() {
    const { drawerPosition } = this.props;
    const rtl = I18nManager.isRTL;
    return rtl
      ? drawerPosition === 'left' ? 'right' : 'left' // invert it
      : drawerPosition;
  }

  componentWillMount() {
    const { openValue } = this.state;

    openValue.addListener(({ value }) => {
      const drawerShown = value > 0;
      const accessibilityViewIsModal = drawerShown;
      if (drawerShown !== this.state.drawerShown) {
        this.setState({ drawerShown, accessibilityViewIsModal });
      }

      if (this.props.keyboardDismissMode === 'on-drag') {
        dismissKeyboard();
      }

      this._lastOpenValue = value;
      if (this.props.onDrawerSlide) {
        this.props.onDrawerSlide({ nativeEvent: { offset: value } });
      }
    });

    this._panResponder = PanResponder.create({
      onMoveShouldSetPanResponder: this._shouldSetPanResponder,
      onPanResponderGrant: this._panResponderGrant,
      onPanResponderMove: this._panResponderMove,
      onPanResponderTerminationRequest: () => false,
      onPanResponderRelease: this._panResponderRelease,
      onPanResponderTerminate: () => {},
    });
  }

  render() {
    const { accessibilityViewIsModal, drawerShown, openValue } = this.state;

    const { drawerBackgroundColor, drawerWidth, drawerPosition } = this.props;

    /**
    * We need to use the "original" drawer position here
    * as RTL turns position left and right on its own
    **/
    const dynamicDrawerStyles = {
      backgroundColor: drawerBackgroundColor,
      width: drawerWidth,
      left: drawerPosition === 'left' ? 0 : null,
      right: drawerPosition === 'right' ? 0 : null,
    };

    /* Drawer styles */
    let outputRange;

    if (this.getDrawerPosition() === 'left') {
      outputRange = [-drawerWidth, 0];
    } else {
      outputRange = [drawerWidth, 0];
    }

    const drawerTranslateX = openValue.interpolate({
      inputRange: [0, 1],
      outputRange,
      extrapolate: 'clamp',
    });
    const animatedDrawerStyles = {
      transform: [{ translateX: drawerTranslateX }],
    };

    /* Overlay styles */
    const overlayOpacity = openValue.interpolate({
      inputRange: [0, 1],
      outputRange: [0, this.props.maxOverlayOpacity],
      extrapolate: 'clamp',
    });
    const animatedOverlayStyles = { opacity: overlayOpacity };
    const pointerEvents = drawerShown ? 'auto' : 'none';


    return (
      <View
        style={{ flex: 1, backgroundColor: 'transparent' }}
        {...this._panResponder.panHandlers}
      >
        <Animated.View style={styles.main}>
          {this.props.children}
        </Animated.View>
        <TouchableWithoutFeedback
          pointerEvents={pointerEvents}
          onPress={this._onOverlayClick}
        >
          <Animated.View
            pointerEvents={pointerEvents}
            style={[styles.overlay, animatedOverlayStyles]}
          />
        </TouchableWithoutFeedback>
        <Animated.View
          accessibilityViewIsModal={accessibilityViewIsModal}
          style={[styles.drawer, dynamicDrawerStyles, animatedDrawerStyles]}
        >
          {this.props.renderNavigationView()}
        </Animated.View>
      </View>
    );
  }

  _onOverlayClick = (e: EventType) => {
    e.stopPropagation();
    if (!this._isLockedClosed() && !this._isLockedOpen()) {
      this.closeDrawer();
    }
  };

  _emitStateChanged = (newState: string) => {
    if (this.props.onDrawerStateChanged) {
      this.props.onDrawerStateChanged(newState);
    }
  };

  openDrawer = (options: DrawerMovementOptionType = {}) => {
    this._emitStateChanged(SETTLING);
    Animated.spring(this.state.openValue, {
        toValue: 1,
        bounciness: 0,
        restSpeedThreshold: 0.1,
        useNativeDriver: this.props.useNativeAnimations,
        ...options,
      })
      .start(() => {
        if (this.props.onDrawerOpen) {
          this.props.onDrawerOpen();
        }
        this._emitStateChanged(IDLE);
      });
  };

  closeDrawer = (options: DrawerMovementOptionType = {}) => {
    this._emitStateChanged(SETTLING);
    Animated.spring(this.state.openValue, {
        toValue: 0,
        bounciness: 0,
        restSpeedThreshold: 1,
        useNativeDriver: this.props.useNativeAnimations,
        ...options,
      })
      .start(() => {
        if (this.props.onDrawerClose) {
          this.props.onDrawerClose();
        }
        this._emitStateChanged(IDLE);
      });
  };

  _handleDrawerOpen = () => {
    if (this.props.onDrawerOpen) {
      this.props.onDrawerOpen();
    }
  };

  _handleDrawerClose = () => {
    if (this.props.onDrawerClose) {
      this.props.onDrawerClose();
    }
  };

  _shouldSetPanResponder = (
    e: EventType,
    { moveX, dx, dy }: PanResponderEventType,
  ) => {
    if (!dx || !dy || Math.abs(dx) < MIN_SWIPE_DISTANCE) {
      return false;
    }

    if (this._isLockedClosed() || this._isLockedOpen()) {
      return false;
    }

    if (this.getDrawerPosition() === 'left') {
      const overlayArea = DEVICE_WIDTH -
        (DEVICE_WIDTH - this.props.drawerWidth);

      if (this._lastOpenValue === 1) {
        if (
          (dx < 0 && Math.abs(dx) > Math.abs(dy) * 3) || moveX > overlayArea
        ) {
          this._isClosing = true;
          this._closingAnchorValue = this._getOpenValueForX(moveX);
          return true;
        }
      } else {
        if (moveX <= 35 && dx > 0) {
          this._isClosing = false;
          return true;
        }

        return false;
      }
    } else {
      const overlayArea = DEVICE_WIDTH - this.props.drawerWidth;

      if (this._lastOpenValue === 1) {
        if (
          (dx > 0 && Math.abs(dx) > Math.abs(dy) * 3) || moveX < overlayArea
        ) {
          this._isClosing = true;
          this._closingAnchorValue = this._getOpenValueForX(moveX);
          return true;
        }
      } else {
        if (moveX >= DEVICE_WIDTH - 35 && dx < 0) {
          this._isClosing = false;
          return true;
        }

        return false;
      }
    }
  };

  _panResponderGrant = () => {
    this._emitStateChanged(DRAGGING);
  };

  _panResponderMove = (e: EventType, { moveX }: PanResponderEventType) => {
    let openValue = this._getOpenValueForX(moveX);

    if (this._isClosing) {
      openValue = 1 - (this._closingAnchorValue - openValue);
    }

    if (openValue > 1) {
      openValue = 1;
    } else if (openValue < 0) {
      openValue = 0;
    }

    this.state.openValue.setValue(openValue);
  };

  _panResponderRelease = (
    e: EventType,
    { moveX, vx }: PanResponderEventType,
  ) => {
    const previouslyOpen = this._isClosing;
    const isWithinVelocityThreshold = vx < VX_MAX && vx > -VX_MAX;

    if (this.getDrawerPosition() === 'left') {
      if (
        (vx > 0 && moveX > THRESHOLD) ||
        vx >= VX_MAX ||
        (isWithinVelocityThreshold && previouslyOpen && moveX > THRESHOLD)
      ) {
        this.openDrawer({ velocity: vx });
      } else if (
        (vx < 0 && moveX < THRESHOLD) ||
        vx < -VX_MAX ||
        (isWithinVelocityThreshold && !previouslyOpen)
      ) {
        this.closeDrawer({ velocity: vx });
      } else if (previouslyOpen) {
        this.openDrawer();
      } else {
        this.closeDrawer();
      }
    } else {
      if (
        (vx < 0 && moveX < THRESHOLD) ||
        vx <= -VX_MAX ||
        (isWithinVelocityThreshold && previouslyOpen && moveX < THRESHOLD)
      ) {
        this.openDrawer({ velocity: (-1) * vx });
      } else if (
        (vx > 0 && moveX > THRESHOLD) ||
        vx > VX_MAX ||
        (isWithinVelocityThreshold && !previouslyOpen)
      ) {
        this.closeDrawer({ velocity: (-1) * vx });
      } else if (previouslyOpen) {
        this.openDrawer();
      } else {
        this.closeDrawer();
      }
    }
  };

  _isLockedClosed = () => {
    return this.props.drawerLockMode === 'locked-closed' &&
      !this.state.drawerShown;
  };

  _isLockedOpen = () => {
    return this.props.drawerLockMode === 'locked-open' &&
      this.state.drawerShown;
  };

  _getOpenValueForX(x: number): number {
    const { drawerWidth } = this.props;

    if (this.getDrawerPosition() === 'left') {
      return x / drawerWidth;
    }

    // position === 'right'
    return (DEVICE_WIDTH - x) / drawerWidth;
  }
}

const styles = StyleSheet.create({
  drawer: {
    position: 'absolute',
    top: 0,
    bottom: 0,
    zIndex: 1001,
  },
  main: {
    flex: 1,
    zIndex: 0,
  },
  overlay: {
    backgroundColor: '#000',
    position: 'absolute',
    top: 0,
    left: 0,
    bottom: 0,
    right: 0,
    zIndex: 1000,
  },
});
