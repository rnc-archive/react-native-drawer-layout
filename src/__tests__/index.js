import { Animated, Text } from 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

import DrawerLayout from '../DrawerLayout';

describe('DrawerLayout', () => {
  it('renders closed correctly', () => {
    const tree = renderer
      .create(
        <DrawerLayout
          drawerWidth={300}
          drawerPosition={DrawerLayout.positions.Left}
          renderNavigationView={() => <Text>Drawer</Text>}
        >
          <Text>Content!</Text>
        </DrawerLayout>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe('lifecycle methods', () => {
    let rendered = null;
    const onDrawerSlide = jest.fn();

    beforeEach(() => {
      rendered = shallow(
        <DrawerLayout
          drawerPosition={DrawerLayout.positions.Left}
          drawerWidth={300}
          onDrawerSlide={onDrawerSlide}
          renderNavigationView={() => <Text>Drawer</Text>}
        >
          <Text>Content!</Text>
        </DrawerLayout>,
      );
    });
    describe('componentWillMount', () => {
      it('should add a listener to the open value', () => {
        const openValue = {
          addListener: jest.fn(),
          interpolate: jest.fn(),
        };
        rendered.setState({ openValue });
        rendered.instance().componentWillMount();
        expect(openValue.addListener).toHaveBeenCalled();
      });

      it('the listener should set the drawerShown value', () => {
        const callbacks = [];
        const openValue = {
          addListener: item => callbacks.push(item),
          interpolate: jest.fn(),
        };
        rendered.setState({ openValue, drawerShown: false });
        rendered.instance().componentWillMount();
        expect(callbacks.length).toBe(1);

        // call callback
        const cb = callbacks[0];
        cb({ value: 0.5 });

        expect(rendered.state('drawerShown')).toBeTruthy();
      });

      it('the listener should call on drawer slide', () => {
        const callbacks = [];
        const openValue = {
          addListener: item => callbacks.push(item),
          interpolate: jest.fn(),
        };
        rendered.setState({ openValue, drawerShown: false });
        rendered.instance().componentWillMount();
        expect(callbacks.length).toBe(1);

        // call callback
        const cb = callbacks[0];
        cb({ value: 0.5 });

        expect(onDrawerSlide).toHaveBeenCalled();
      });

      it('the listener should dismiss the keyboard');
    });
  });

  describe('pan responder', () => {
    describe('_getOpenValueForX', () => {
      it('should return 1 if the drawer is fully opened');
      it('should return 0 if the drawer is fully closed');
      it('should work for left and right');
    });

    describe('panResponderRelease', () => {
      it('should open the drawer after threshold for left');
      it('should open the drawer according to speed for left');
      it('should close the drawer after threshold for left');
      it('should close the drawer according to speed for left');
      it('should open the drawer after threshold for right');
      it('should open the drawer according to speed for right');
      it('should close the drawer after threshold for right');
      it('should close the drawer according to speed for right');
    });
    describe('panResponderMove', () => {
      it('should set the open value');
      it('should set the open value to max 1');
      it('should set the open value to min 0');
    });
    describe('panResponderGrant', () => {
      it('should emit dragging');
    });
    describe('shouldSetPanResponder', () => {
      it('returns false if locked');
      describe('open', () => {
        it('returns true if significant swipe to right, position left');
        it('returns true if significant swipe to left, position right');
      });

      describe('closed', () => {
        it('returns true if significant sw to right, position right');
        it('returns true if significant swipe to left, position left');
      });

      it('returns false if vertical swipe');
    });
  });

  describe('external methods', () => {
    let rendered = null;
    let onDrawerStateChanged = null;

    beforeEach(() => {
      onDrawerStateChanged = jest.fn();
      onDrawerOpen = jest.fn();

      rendered = shallow(
        <DrawerLayout
          drawerPosition={DrawerLayout.positions.Left}
          drawerWidth={300}
          onDrawerStateChanged={onDrawerStateChanged}
          renderNavigationView={() => <Text>Drawer</Text>}
        >
          <Text>Content!</Text>
        </DrawerLayout>,
      );
    });

    describe('openDrawer', () => {
      it('should emit event', () => {
        rendered.instance().openDrawer();
        expect(onDrawerStateChanged).toHaveBeenCalledWith('Settling');
      });

      it('should start animation');
      it('should emit idle event when done with animation');
      it('should call callback when done with animation');
    });

    describe('closeDrawer', () => {
      it('should emit event', () => {
        rendered.instance().closeDrawer();
        expect(onDrawerStateChanged).toHaveBeenCalledWith('Settling');
      });

      it('should start animation');
      it('should emit idle event when done with animation');
      it('should call callback when done with animation');
    });
  });
});
