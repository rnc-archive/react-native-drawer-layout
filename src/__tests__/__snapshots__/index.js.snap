exports[`DrawerLayout renders closed correctly 1`] = `
<View
  onMoveShouldSetResponder={[Function]}
  onMoveShouldSetResponderCapture={[Function]}
  onResponderEnd={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderReject={[Function]}
  onResponderRelease={[Function]}
  onResponderStart={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  onStartShouldSetResponderCapture={[Function]}
  style={
    Object {
      "backgroundColor": "transparent",
      "flex": 1,
    }
  }>
  <View
    style={
      Object {
        "flex": 1,
      }
    }>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail">
      Content!
    </Text>
  </View>
  <View
    accessibilityComponentType={undefined}
    accessibilityLabel={undefined}
    accessibilityTraits={undefined}
    accessible={true}
    hitSlop={undefined}
    onLayout={undefined}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    pointerEvents="none"
    style={
      Object {
        "backgroundColor": "#000",
        "bottom": 0,
        "left": 0,
        "opacity": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
        "zIndex": 1000,
      }
    }
    testID={undefined} />
  <View
    style={
      Object {
        "backgroundColor": undefined,
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "top": 0,
        "transform": Array [
          Object {
            "translateX": -300,
          },
        ],
        "width": 300,
        "zIndex": 1001,
      }
    }>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail">
      Drawer
    </Text>
  </View>
</View>
`;
